caikit/runtime/http_server.py

```diff
@@ -0,0 +1,460 @@
+        unvicorn_timeout_graceful_shutdown = get_config().runtime.http.timeout
+
+        self.app = FastAPI()
+
+        # Set up the central predict servicer
+        inference_service = ServicePackageFactory().get_service_package(
+            ServicePackageFactory.ServiceType.INFERENCE,
+        )
+        self.global_predict_servicer = GlobalPredictServicer(inference_service)
+
+        # Set up the central train servicer
+        # TODO: uncomment later on
+        # train_service = ServicePackageFactory().get_service_package(
+        #     ServicePackageFactory.ServiceType.TRAINING,
+        # )
+        # self.global_train_servicer = GlobalTrainServicer(train_service)
+        self.package_name = inference_service.descriptor.full_name.rsplit(".", 1)[0]
+
+        # Bind all routes to the server
+        self._bind_routes(inference_service)
+        # self._bind_routes(train_service)
+
+        # Parse TLS configuration
+        tls_kwargs = {}
+        if (
+            self.tls_config
+            and self.tls_config.server.key
+            and self.tls_config.server.cert
+        ):
+            log.info("<RUN10001905I>", "Running with TLS")
+            tls_kwargs["ssl_keyfile"] = self.tls_config.server.key
+            tls_kwargs["ssl_certfile"] = self.tls_config.server.cert
+            if self.tls_config.client.cert:
+                log.info("<RUN10001809I>", "Running with mutual TLS")
+                tls_kwargs["ssl_ca_certs"] = self.tls_config.client.cert
+                tls_kwargs["ssl_cert_reqs"] = ssl.CERT_REQUIRED
+
+        # Start the server and run forever
+        config = uvicorn.Config(
+            self.app,
+            host="0.0.0.0",
+            port=self.port,
+            log_level=None,
+            log_config=None,
+            timeout_graceful_shutdown=unvicorn_timeout_graceful_shutdown,
+            **tls_kwargs,
+        )
+        self.server = uvicorn.Server(config=config)
+        self.uvicorn_server_thread = threading.Thread(target=self.server.run)
+
+    def __del__(self):
+        if get_config().runtime.metering.enabled:
+            self._stop_metering()
+
+    # Override context manager impl
+    def __enter__(self):
+        self.start()
+        return self
+
+    # TODO: what should actually happen here?
+    def __exit__(self, type_, value, traceback):
+        pass
+
+    def start(self):
+        """Start the server (blocking)"""
+        self.server.run()
+
+    @contextmanager
+    def run_in_thread(self):
+        self.uvicorn_server_thread.start()
+        try:
+            while not self.server.started:
+                time.sleep(1e-3)
+            log.info("HTTP Server is running in thread")
+            yield
+
+        finally:
+            log.info("Shutting down HTTP Server")
+            self.server.should_exit = True
+            self.uvicorn_server_thread.join()
+
+    ##########
+    ## Impl ##
+    ##########
+    def _stop_metering(self):
```

Comment:
Very nice! I wonder if this should be a function on `global_predict_servicer` itself and if it should be called in the grpc server too